src/core/utils/DenoiserManager.js

```diff
@@ -0,0 +1,208 @@
+
+import { NoBlending } from 'three';
+import * as THREE from 'three';
+
+export class DenoiserManager {
+
+	constructor( renderer ) {
+
+		this.renderer = renderer;
+		this.denoiser = new Denoiser( 'webgl', renderer.domElement );
+		this.denoiser.inputMode = 'webgl';
+		this.denoiser.outputMode = 'webgl';
+		this.denoiser.weightsUrl = 'https://cdn.jsdelivr.net/npm/denoiser/tzas';
+
+		//this.denoiser.debugging = true;
+		//this.denoiser.usePassThrough = true;
+		//this.rawCanvas = document.getElementById( 'rawCanvas' );
+		//this.denoiser.setCanvas( this.rawCanvas );
+
+		this.isDenoising = false;
+		this.fadeTime = 500;
+		this.denoiserFinished = 0;
+
+		// Same as pathtracer so tonemapping is the same
+		this.ptMaterial = new ClampedInterpolationMaterial( {
+			map: null,
+			transparent: true,
+			blending: NoBlending,
+
+			premultipliedAlpha: renderer.getContextAttributes().premultipliedAlpha,
+		} );
+		this.ptMaterial.opacity = 1;
+
+		// Material to blend between pathtracer and denoiser
+		// TODO: Not sure if  params are required
+		this.blendMaterial = new BlendMaterial();
+
+		this.quad = new FullScreenQuad( this.ptMaterial );
+		this.createConversionRenderTarget( renderer.domElement.width, renderer.domElement.height );
+
+	}
+
+	async denoise( rawPathtracedTexture, albedoTexture, normalTexture ) {
+
+		this.isDenoising = true;
+		// Adjust the height /width if changed from before
+		const height = this.renderer.domElement.height;
+		const width = this.renderer.domElement.width;
+
+		if ( this.denoiser.height !== height || this.denoiser.width !== width ) {
+
+			this.denoiser.width = width;
+			this.denoiser.height = height;
+			this.createConversionRenderTarget( width, height );
+
+			/* Used when debugging
+			if ( this.rawCanvas ) {
+
+				this.rawCanvas.width = width;
+				this.rawCanvas.height = height;
+
+			}*/
+
+		}
+
+		// set so we can access later when blending
+		this.pathtracedTexture = this.getCorrectPathtracerTexture( rawPathtracedTexture );
+		const colorWebGLTexture = this.getWebGLTexture( this.pathtracedTexture );
+		const albedoWebGLTexture = this.getWebGLTexture( albedoTexture );
+		const normalWebGLTexture = this.getWebGLTexture( normalTexture );
+
+		//* run the denoiser
+		this.renderer.resetState();
+
+		/* The setting of inputs is async and running too quick before execute. this will be fixed in a future version
+		see https://github.com/DennisSmolek/Denoiser/issues/21
+		const denoisedWebGLTexture = await this.denoiser.execute( colorWebGLTexture, albedoWebGLTexture, normalWebGLTexture );
+		*/
+
+		// Set input Textures with await
+		await this.denoiser.setInputTexture( 'color', colorWebGLTexture );
+		if ( albedoTexture ) await this.denoiser.setInputTexture( 'albedo', albedoWebGLTexture );
+		if ( normalTexture ) await this.denoiser.setInputTexture( 'normal', normalWebGLTexture );
+		// Run the denoiser
+		const denoisedWebGLTexture = await this.denoiser.execute();
```

Comment:
I see some logs about tiling happening - are the tiles being rendered over multiple frames? It would be nice if the page didn't lock up quite so much when denoising is happening. The tiles could also be used to display a completion percentage.

Reply:
The tile generation we could probably wire up to wait for a frame before going to the next tile (allowing less lockup) but I wonder if this would slow things down on higher speed machines.[Tensorflow has a method for this](https://js.tensorflow.org/api/latest/#nextFrame)

Tile reconstruction (which is the second half) should happen as fast as possible, I'm not sure we want to divide that up.

I'll log down both strategies to try.

One thing to remember, With the WebGL backend the first execution is SIGNIFICANTLY longer than the others. Tensorflow waits to compile shaders until you actually execute.It's suggested we run a few images through to "prime" the model but this would cause it to be slow at load time which I was avoiding.

I could see a "ready()" callback that fires after a warm up image is passed.

Reply:
> With the WebGL backend the first execution is SIGNIFICANTLY longer than the others. Tensorflow waits to compile shaders until you actually execute

Is there any way to use the [KHR_parallel_shader_compile](https://developer.mozilla.org/en-US/docs/Web/API/KHR_parallel_shader_compile) extension so the shaders will compile asynchronously? Three.js implements this with [WebGLRenderer.compileAsync](https://threejs.org/docs/?q=webglrenderer#api/en/renderers/WebGLRenderer.compileAsync), which the PathTracer uses [here](https://github.com/gkjohnson/three-gpu-pathtracer/blob/d5b1f148afa82d6f9a28be4870e8d9fb664e593c/src/core/PathTracingRenderer.js#L274) so the page does not block while the path tracing shader compiles.